File: test/cli.test.mjs
/**
 * End-to-end checks against the real entry points.
 *
 * The unit tests exercise `renderHud` directly, which cannot catch a launcher
 * pointing at the wrong file or a flag that throws before rendering — and both
 * of those show up as an empty statusline, the one failure mode a statusline
 * must never have.
 */

import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { after, describe, it } from "node:test";

const here = dirname(fileURLToPath(import.meta.url));
const root = join(here, "..");
const entry = join(root, "bin", "claude-readout.mjs");
const launcher = join(root, "bin", "claude-readout.sh");
const fixture = readFileSync(join(here, "fixture-session.json"), "utf8");

// An isolated HOME so the suite never reads or writes the developer's own
// config, cache or credentials.
const sandbox = mkdtempSync(join(tmpdir(), "claude-readout-cli-"));
after(() => rmSync(sandbox, { recursive: true, force: true }));

const env = {
  PATH: process.env.PATH,
  HOME: sandbox,
  XDG_CONFIG_HOME: join(sandbox, "config"),
  XDG_CACHE_HOME: join(sandbox, "cache"),
  CLAUDE_CONFIG_DIR: join(sandbox, "claude"), // no credentials -> no usage call
  NO_COLOR: "1",
};

function run(command, args, input = "") {
  return execFileSync(command, args, { input, env, encoding: "utf8", timeout: 15_000 });
}

describe("cli", () => {
  it("renders from a statusline payload", () => {
    const out = run(process.execPath, [entry], fixture);
    assert.match(out, /Opus 5/);
    assert.match(out, /5h .*38%/);
    assert.equal(out.endsWith("\n"), true);
  });

  it("still prints a line when stdin is empty or garbage", () => {
    assert.equal(typeof run(process.execPath, [entry], ""), "string");
    assert.equal(typeof run(process.execPath, [entry], "not json at all"), "string");
  });

  it("runs through the shell launcher", () => {
    // Regression: the launcher hard-codes its entry filename, so a rename can
    // silently point it at a file that no longer exists.
    const out = run("sh", [launcher], fixture);
    assert.match(out, /Opus 5/);
  });

  it("finds node even with no PATH inherited", () => {
    const out = execFileSync("sh", [launcher], {
      input: fixture,
      env: { HOME: sandbox, NO_COLOR: "1" },
      encoding: "utf8",
      timeout: 15_000,
    });
    assert.match(out, /Opus 5/);
  });

  it("prints the legend, the ramp and the doctor report", () => {
    assert.match(run(process.execPath, [entry, "--legend"]), /weekly usage across all models/);
    assert.match(run(process.execPath, [entry, "--ramp"]), /100%/);
    assert.match(run(process.execPath, [entry, "--doctor"]), /oauth token:/);
  });

  it("honours NO_COLOR", () => {
    const out = run(process.execPath, [entry], fixture);
    assert.equal(/\x1b\[/.test(out), false);
  });
});

File: cli_test.go
package main

import (
	"bytes"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
)

var cliBinary string

func TestMain(m *testing.M) {
	dir, err := os.MkdirTemp("", "claude-readout-cli-")
	if err != nil {
		panic(err)
	}
	cliBinary = filepath.Join(dir, "claude-readout")
	cmd := exec.Command("go", "build", "-o", cliBinary, ".")
	if out, buildErr := cmd.CombinedOutput(); buildErr != nil {
		os.Stderr.Write(out)
		os.RemoveAll(dir)
		os.Exit(1)
	}
	code := m.Run()
	os.RemoveAll(dir)
	os.Exit(code)
}

func runCLI(t *testing.T, input string, args ...string) string {
	t.Helper()
	root := t.TempDir()
	cmd := exec.Command(cliBinary, args...)
	cmd.Stdin = strings.NewReader(input)
	cmd.Env = []string{
		"HOME=" + root,
		"XDG_CONFIG_HOME=" + filepath.Join(root, "config"),
		"XDG_CACHE_HOME=" + filepath.Join(root, "cache"),
		"CLAUDE_CONFIG_DIR=" + filepath.Join(root, "claude"),
		"NO_COLOR=1",
		"PATH=" + os.Getenv("PATH"),
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("claude-readout %v: %v\n%s", args, err, out)
	}
	return string(out)
}

func TestCLIRendersFixture(t *testing.T) {
	fixture, err := os.ReadFile(filepath.Join("testdata", "fixture-session.json"))
	if err != nil {
		t.Fatal(err)
	}
	out := runCLI(t, string(fixture))
	if !strings.Contains(out, "Opus 5") || !regexp.MustCompile(`5h .*38%`).MatchString(out) || !strings.HasSuffix(out, "\n") {
		t.Fatalf("unexpected output: %q", out)
	}
}

func TestCLIEmptyAndGarbageStillPrintALine(t *testing.T) {
	for _, input := range []string{"", "not json at all"} {
		out := runCLI(t, input)
		if out == "" || !strings.HasSuffix(out, "\n") {
			t.Fatalf("input %q produced %q", input, out)
		}
	}
}

func TestCLILegend(t *testing.T) {
	if out := runCLI(t, "", "--legend"); !strings.Contains(out, "weekly usage across all models") {
		t.Fatalf("legend output: %q", out)
	}
}

func TestCLIRamp(t *testing.T) {
	if out := runCLI(t, "", "--ramp"); !strings.Contains(out, "100%") {
		t.Fatalf("ramp output: %q", out)
	}
}

func TestCLIDoctor(t *testing.T) {
	if out := runCLI(t, "", "--doctor"); !strings.Contains(out, "oauth token:") {
		t.Fatalf("doctor output: %q", out)
	}
}

func TestCLINoColorHasNoEscapeByte(t *testing.T) {
	fixture, err := os.ReadFile(filepath.Join("testdata", "fixture-session.json"))
	if err != nil {
		t.Fatal(err)
	}
	out := runCLI(t, string(fixture))
	if bytes.ContainsRune([]byte(out), '\x1b') {
		t.Fatalf("output contains escape byte: %q", out)
	}
}
